index.src.html

```diff
@@ -699,6 +757,53 @@ <h3 id="determine-requests-referrer">
         <dt><a>"<code>unsafe-url</code>"</a></dt>
         <dd>Return <var>referrerURL</var>.</dd>
 
+        <dt><a>"<code>safe-origin</code>"</a></dt>
+        <dd>
+          <ol>
+            <li>
+              If |environment| is not null:
+              <ol>
+                <li>
+                  If |environment| is <a>TLS-protected</a> <em>and</em>
+                  the <a>origin</a>
+                  of <var>request</var>'s <a href="https://fetch.spec.whatwg.org/#concept-request-current-url">current
+                  URL</a> is not an <a><em>a priori</em> authenticated
+                  URL</a>, then return <code>no referrer</code>.
+                </li>
+              </ol>
+            </li>
+            <li>Return |referrerOrigin|.
+          </ol>
+        </dd>
+
+        <dt><a>"<code>safe-origin-when-cross-origin</code>"</a></dt>
+        <dt>the empty string</dt>
+        <dd>
+          <ol>
+            <li>
```

Comment:
If I'm reading this correctly, the algorithm contained in this patch is:

```
if httpsPage and httpResource:
    if page and resource are same-origin:
        return fullURL
    else:
        return origin
else:
    return no-referrer
```

That's not quite right, I would suggest this instead:

```
1. If request is a same-origin request, then return referrerURL.
2. If environment is not null:
    1. If environment is TLS-protected and the origin of request’s current
       URL is not an a priori authenticated URL, then return no referrer.
3. Otherwise, return referrerOrigin.
```